### ISSUE TITLE
Is C# 11.0 supported? Invalid option '11.0' for /langversion.

### ISSUE THREAD

**Issue by RudeySH** [aspnet/RoslynCodeDomProvider#152]
Is C# 11.0 supported? I wasn't able to get this to work with the following configuration:

```
  <system.codedom>
    <compilers>
      <compiler extension=".cs" language="c#;cs;csharp" warningLevel="4"
        compilerOptions="/langversion:11.0"
        type="Microsoft.CodeDom.Providers.DotNetCompilerPlatform.CSharpCodeProvider, Microsoft.CodeDom.Providers.DotNetCompilerPlatform, Version=4.1.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35"/>
      <compiler extension=".vb" language="vb;vbs;visualbasic;vbscript" warningLevel="4"
        compilerOptions="/langversion:11.0 /define:_MYTYPE=\&quot;Web\&quot; /optionInfer+"
        type="Microsoft.CodeDom.Providers.DotNetCompilerPlatform.VBCodeProvider, Microsoft.CodeDom.Providers.DotNetCompilerPlatform, Version=4.1.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35"/>
    </compilers>
  </system.codedom>
```

Compiler Error Message: CS1617: Invalid option '11.0' for /langversion. Use '/langversion:?' to list supported values.

I also tried '11', but that also doesn't work.

**Comment by precisioninfinity** [comment]
I'm pretty sure nothing after C-sharp 7.3 is supported. I read once somewhere that you can get C-sharp 8 going by doing some hacky workaround but that would be a bit of a risk. Since that has worked for people in the past there is a possibility that you could get greater than C sharp 8 by doing the same thing. You could search around online for that post.

**Comment by CZEMacLeod** [comment]
If you include the package `Microsoft.Net.Compilers.Toolset` with GeneratePathProperty set to true, you can then set the property `RoslynToolPath` to `$(PkgMicrosoft_Net_Compilers_Toolset)\tasks\net472`
Then you can use langversion up to the matching the version of `Microsoft.Net.Compilers.Toolset` you select.
See [Roslyn Nuget Packages Versions](https://github.com/dotnet/roslyn/blob/main/docs/wiki/NuGet-packages.md#versioning)
I don't know if C# 11 is actually available yet (might be in 4.5.0), but C# 10 definitely is. Also remember that the vb langversion flag doesn't follow the c# one, and that some new c# features aren't available in .Net Framework.
A lot of this hookup work is already done if you use the SDK style project available at [MSBuild.SDK.SystemWeb](https://github.com/CZEMacLeod/MSBuild.SDK.SystemWeb) (Full disclosure - I maintain that repo).
I also think that V4.1.0 of this package should include the Roslyn compilers version 4.1.0 by default, and that should support C# 10 anyway.

**Comment by saucecontrol** [comment]
> I also think that V4.1.0 of this package should include the Roslyn compilers version 4.1.0 by default, and that should support C# 10 anyway.

That's correct.  It's unfortunate that the first release of this package in 3 years is using a year-old Roslyn version instead of current.

Unfortunately, the workaround using the `RoslynToolPath` doesn't work with Web Site solutions (no project file to put it in), so the only option to get C# 11 support is to script or manually copy the `roslyn` folder contents from the current [Toolset](https://www.nuget.org/packages/Microsoft.Net.Compilers.Toolset/) package to the app's `bin` folder.

**Comment by RudeySH** [comment]
> I'm pretty sure nothing after C-sharp 7.3 is supported.

Up to C# 10 works without requiring the need for a workaround.

I'll give @CZEMacLeod's workaround a try to see if C# 11 is available.

**Comment by precisioninfinity** [comment]
> > I'm pretty sure nothing after C-sharp 7.3 is supported.
> 
> Up to C# 10 works without requiring the need for a workaround.
> 
> I'll give @CZEMacLeod's workaround a try to see if C# 11 is available.

That's cool, good to know, thanks! If you find out I'd be curious if you get C# 11 going too.

**Comment by StephenMolloy** [comment]
As others have mentioned, C# only officially supports the full .Net Framework up to version 7.3. As a package offered by Microsoft, we can't officially endorse setting the /langversion to anything higher than that. (Of course, it is possible if you want to blaze that trail yourself. We just can't do it for you.)

Also as others have mentioned, Roslyn 4.1 only supports C# 10, so 11 is not an option with the 4.1 toolset. You can of course "bring your own toolset" with the `RoslynToolPath` approach if you don't want to wait for a version of this package to ship with an updated Roslyn tool set.

Finally, generally speaking, this is not a package that will be rev-ving frequently. Full .Net and ASP.Net are mostly in maintenance mode and do not iterate frequently anymore. It is an unfortunate architectural effect of VS and ASP.Net compilation that we need to directly include the Roslyn tools instead of just using a nuget dependency. I envision that there will be periodic refreshes of this package just to update the included Roslyn tool set. But those refreshes won't come at the same pace as the Roslyn tools themselves. For users who need that more rapid cadence, `RoslynToolPath` is the way to go for that fine-grained control.

(The 4.1 release here was a large change in the mechanics of this package. We want to allow some bake time as an "rtm" package with "rtm" visibility to shake out any final issues we might have with the 3.X->4.X jump before we start our slow update cadence - as we don't expect/want the update cadence to do anything more than just update the Roslyn tools so updates are a low-friction exercise for people.)
